rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,540 @@
+overheads.  This will introduce some overhead for brgemm objects
+creation though (potential extra copies upon brgemm constructor call).
+
+#### Other considerations
+
+The recommendation as a starting point is to make query mandatory: no
+layout forcing on user side. This simplifies the configurations
+supported by brgemm. Given that we have no support internally for
+transa/tranb, those will not be exposed as well. We will also support
+only row major representation, and not expose a knob to select
+row/column major.
+
+Given the small number of layouts, we will expose a layout tag enum
+(next section will address the lack of need for metadata handling by
+oneDNN brgemm).
+
+## Low precision handling.
+
+The API will be limited to support downconversion/upconversion before
+computation in a similar way that
+[dnnl::fpmath_mode](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_fpmath_mode.html#)
+provides.
+
+If a user wants to implement quantized computation with zero-point,
+they will have to precompute compensation term and pass it to brgemm
+computation as bias.
+
+The main drivers behind this choice are:
+- flexibility on the user side, as any custom quantization scheme can
+  be handled. It removes the dependency on oneDNN wrt when/how any
+  compensation term should be computed.
+- consistency on the user side. Whether they implement a custom
+  quantization scheme or a standard one, the programming model is the
+  same.
+- supporting metadata on oneDNN sides would force using higher level
+  abstraction that can impact performance. Also, there is little to no
+  performance gain to expect from this level of fusion.
+
+## Hardware specific discussions
+
+### ISA selection
+
+Here two options: 
+- the user can specify for which ISA the kernel can be generated,
+- oneDNN does it transparently under the hood to generate best kernel
+  for current platform.
+
+Note that current oneDNN implementation just takes ISA as a maximum
+ISA.  To simplify the programming model, isa selection will not appear
+in block level APIs. However, `dnnl::set_max_cpu_isa` will still be
+effective, and can be used by the user to control the maximum isa
+level used by block level APIs.
+
+
+### Handling of architectural state 
+
+Matrix accelerators typically require to set an architectural state
+(e.g. Intel AMX or ARM SME).
+
+A few elements about those architectural states:
+- there is a per-core setting, which is typically costly in terms of
+  cycle count.
+- this state is block size dependent, so executing two brgemm kernels
+  with different shapes require an expensive reconfiguration between
+  each.
+- when matrix accelerator is armed, the max frequency for the core is
+  altered, so performance of sequential/vector code can be impacted.
+- there are some OS syscall to do in order to enable expanded stack
+  space.
+
+
+There are two options here:
+- expose two APIs, `brgemm::set_hw_context()` and
```

Comment:
If we need to run two brgemm kernels, do we need to run `brgemm::release_hw_context()` and `brgemm::set_hw_context()` in between? Can we merge them into `brgemm:reset_hw_context()` to reduce the overhead?

Reply:
There is a comment on API level that no release is needed between two calls to `set_hw_context()`.
However adding resest_hw_context is cleaner, so adding that to the proposal. Thanks.